x-pack/heartbeat/monitors/browser/source/zipurl.go

```diff
@@ -127,24 +123,35 @@ func unzip(tf *os.File, targetDir string, folder string) error {
 	return nil
 }
 
+// unzip file takes a given directory and a zipped file and extracts
+// all the contents of the file based on the provided folder path,
+// if the folder path is empty, it extracts the contents based on file
+// tree structure
 func unzipFile(workdir string, folder string, f *zip.File) error {
-	folderPaths := strings.Split(folder, string(filepath.Separator))
-	var folderDepth = 1
-	for _, path := range folderPaths {
-		if path != "" {
-			folderDepth++
+	var destPath string
+	if folder != "" {
+		folderPaths := strings.Split(folder, string(filepath.Separator))
+		var folderDepth = 1
+		for _, path := range folderPaths {
+			if path != "" {
+				folderDepth++
+			}
 		}
-	}
-	splitZipFileName := strings.Split(f.Name, string(filepath.Separator))
-	root := splitZipFileName[0]
+		splitZipFileName := strings.Split(f.Name, string(filepath.Separator))
+		root := splitZipFileName[0]
 
-	prefix := filepath.Join(root, folder)
-	if !strings.HasPrefix(f.Name, prefix) {
-		return nil
-	}
+		prefix := filepath.Join(root, folder)
+		if !strings.HasPrefix(f.Name, prefix) {
+			return nil
+		}
 
-	sansFolder := splitZipFileName[folderDepth:]
-	destPath := filepath.Join(workdir, filepath.Join(sansFolder...))
+		sansFolder := splitZipFileName[folderDepth:]
+		destPath = filepath.Join(workdir, filepath.Join(sansFolder...))
+	} else {
+		//nolint:gosec // zip slip vulnerability, but we control the zipping when
+		//pushing the monitors
+		destPath = filepath.Join(workdir, f.Name)
```

Comment:
Although we control the zipping when pushing monitors I'd expect that an attacker could still push malicious `zip` files and compromise all hosts with Heartbeat if they end up having the secrets for Kibana, so I think it would be worth doing some sanitisation here. What do you think?

It seems like all we'd have to do is:

1. Get the shortest possible path with `filepath.Clean(filepath.Join(workdir, f.Name)) + string(os.PathSeparator)`
2. Check if the prefix of `filepath.Join(workdir, f.Name)` is the same as the shortest possible path

Reply:
Great point on this, will add some sanitization logic to catch the behaviour 👍🏽

Reply:
Done.